### PR TITLE
[Merged by Bors] - feat: port Analysis.Convolution

### PR DESCRIPTION
**Pull request by urkud**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Ruben-VandeVelde** [comment]
Ah, I wasn't paying attention and also started on this. Only got as far as the second theorem, though:
```diff
--- Mathlib/Analysis/Convolution.lean
+++ Mathlib/Analysis/Convolution.lean
@@ -123,13 +123,17 @@ variable [AddGroup G] [TopologicalSpace G]
 theorem convolution_integrand_bound_right_of_le_of_subset {C : ℝ} (hC : ∀ i, ‖g i‖ ≤ C) {x t : G}
     {s u : Set G} (hx : x ∈ s) (hu : -tsupport g + s ⊆ u) :
     ‖L (f t) (g (x - t))‖ ≤ u.indicator (fun t => ‖L‖ * ‖f t‖ * C) t := by
-  refine' le_indicator (fun t ht => _) (fun t ht => _) t
+  -- Porting note: added `suffices`
+  suffices (fun t => ‖(L (f t)) (g (x - t))‖) ≤ indicator u (fun t ↦ ‖L‖ * ‖f t‖ * C) by
+    apply this t
+  refine' le_indicator (fun t _ => _) (fun t ht => _)
   · refine' (L.le_op_norm₂ _ _).trans _
-    apply mul_le_mul_of_nonneg_left (hC _) (mul_nonneg (norm_nonneg _) (norm_nonneg _))
+    apply mul_le_mul_of_nonneg_left (hC _) (mul_nonneg (norm_nonneg L) (norm_nonneg _))
   · have : x - t ∉ support g := by
       refine' mt (fun hxt => _) ht
       apply hu
-      refine' ⟨_, _, set.neg_mem_neg.mpr (subset_closure hxt), hx, _⟩
+      refine' ⟨_, _, Set.neg_mem_neg.mpr (subset_closure hxt), hx, _⟩
+      simp only
       rw [neg_sub, sub_add_cancel]
     rw [nmem_support.mp this, (L _).map_zero, norm_zero]
 #align convolution_integrand_bound_right_of_le_of_subset convolution_integrand_bound_right_of_le_of_subset
@@ -137,8 +141,8 @@ theorem convolution_integrand_bound_right_of_le_of_subset {C : ℝ} (hC : ∀ i,
 theorem HasCompactSupport.convolution_integrand_bound_right_of_subset (hcg : HasCompactSupport g)
     (hg : Continuous g) {x t : G} {s u : Set G} (hx : x ∈ s) (hu : -tsupport g + s ⊆ u) :
     ‖L (f t) (g (x - t))‖ ≤ u.indicator (fun t => ‖L‖ * ‖f t‖ * ⨆ i, ‖g i‖) t := by
-  apply convolution_integrand_bound_right_of_le_of_subset _ (fun i => _) hx hu
-  exact le_ciSup (hg.norm.bdd_above_range_of_has_compact_support hcg.norm) _
+  refine convolution_integrand_bound_right_of_le_of_subset _ (fun i => ?_) hx hu
+  exact le_ciSup (hg.norm.bddAbove_range_of_hasCompactSupport hcg.norm) _
 #align has_compact_support.convolution_integrand_bound_right_of_subset HasCompactSupport.convolution_integrand_bound_right_of_subset
 
 theorem HasCompactSupport.convolution_integrand_bound_right (hcg : HasCompactSupport g)
```

**Comment by urkud** [comment]
@Ruben-VandeVelde After a few merge conflicts, I started pushing to github & opening a PR right after running `start_port.sh`.

File: Mathlib/Analysis/Convolution.lean [hunk 1]
```diff
@@ -0,0 +1,1677 @@
...
+    (hgk : ∀ᵐ x ∂ν, ConvolutionExistsAt (fun x => ‖g x‖) (fun x => ‖k x‖) x (mul ℝ ℝ) μ)
+    (hfgk :
+      ConvolutionExistsAt (fun x => ‖f x‖) ((fun x => ‖g x‖) ⋆[mul ℝ ℝ, μ] fun x => ‖k x‖) x₀
+        (mul ℝ ℝ) ν) :
+    ((f ⋆[L, ν] g) ⋆[L₂, μ] k) x₀ = (f ⋆[L₃, ν] g ⋆[L₄, μ] k) x₀ := by
+  refine' convolution_assoc' L L₂ L₃ L₄ hL hfg (hgk.mono fun x hx => hx.ofNorm L₄ hg hk) _
+  -- the following is similar to `integrable.convolution_integrand`
+  have h_meas :
+    AEStronglyMeasurable (uncurry fun x y => L₃ (f y) (L₄ (g x) (k (x₀ - y - x)))) (μ.prod ν) :=
+    by
```
Comment (urkud) [inline Mathlib/Analysis/Convolution.lean:1128]:
```suggestion
    AEStronglyMeasurable (uncurry fun x y => L₃ (f y) (L₄ (g x) (k (x₀ - y - x))))
      (μ.prod ν) := by
```

**Comment by fpvandoorn** [comment]
This file is incredibly slow since [`MeasureTheory.Measure`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/MeasureSpaceDef.html#MeasureTheory.Measure) is overloaded with [`Measure`](https://leanprover-community.github.io/mathlib4_docs/Init/WF.html#Measure), that alone probably adds multiple minutes of compile time in this file.

**Comment by urkud** [comment]
What should we do? Introduce `local notation "Measure" := MeasureTheory.Measure`? Something else? Ask core devs to add a namespace to `_root_.Measure`? (is it an option?)

**Comment by fpvandoorn** [comment]
I started https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Measure.20is.20overloaded/near/366584965

**Comment by fpvandoorn** [comment]
`local notation "Measure" => MeasureTheory.Measure` would be nice, but it doesn't work. It still considers `_root_.Measure`. I think as a workaround I'll just search-and-replace `Measure` to `MeasureTheory.Measure` (with a porting note).

**Review by fpvandoorn — COMMENTED** [review]
I didn't finish the review yet, but I pushed some changes. The `MeasureTheory.Measure` fix speeds the file up by much more than 2x.

File: Mathlib/Analysis/NormedSpace/OperatorNorm.lean [hunk 2]
```diff
@@ -611,6 +615,12 @@ theorem le_op_norm₂ [RingHomIsometric σ₁₃] (f : E →SL[σ₁₃] F →SL
   (f x).le_of_op_norm_le (f.le_op_norm x) y
 #align continuous_linear_map.le_op_norm₂ ContinuousLinearMap.le_op_norm₂
 
+-- porting note: new theorem
+theorem le_op_norm₂_of_le₃ [RingHomIsometric σ₁₃] (f : E →SL[σ₁₃] F →SL[σ₂₃] G) {x : E} {y : F}
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/NormedSpace/OperatorNorm.lean:619]: I agree about the name. The note says that we didn't loose `#align`. Someone asked me to add a note like this when I was porting another file, so I started adding them.

File: Mathlib/Analysis/Convolution.lean [hunk 3]
```diff
@@ -0,0 +1,1681 @@
...
+  -- was refine'
+  apply hasFDerivAt_integral_of_dominated_of_fderiv_le zero_lt_one h1 _ (h2 x₀)
+  · refine' eventually_of_forall fun t x hx =>
+      (hcg.fderiv 𝕜).convolution_integrand_bound_right L' (hg.continuous_fderiv le_rfl)
+        (ball_subset_closedBall hx)
+  · rw [integrable_indicator_iff hK'.measurableSet]
+    exact ((hf.integrableOn_isCompact hK').norm.const_mul _).mul_const _
+  · exact eventually_of_forall fun t x _ => (L _).hasFDerivAt.comp x (h3 x t)
+  · exact hcg.convolutionExists_right L hf hg.continuous x₀
+-- · exact K'.indicator fun t => ‖L'‖ * ‖f t‖ * ⨆ x, ‖fderiv 𝕜 g x‖
```
Comment (Parcly-Taxel) [inline Mathlib/Analysis/Convolution.lean:1211]:
```suggestion
```

**Review by j-loreaux — COMMENTED** [review]
Looks mostly good to me aside from the comments below (and Floris' naming suggestion). Is the file fast enough now with the `MeasureTheory.Measure` change?

File: Mathlib/Analysis/Convolution.lean [hunk 4]
```diff
@@ -1542,68 +1480,68 @@ theorem contDiffOn_convolution_right_with_param {f : G → E} {n : ℕ∞} (L :
...
-  let eG : Type max uG uE' uF uP := ULift G
-  borelize eG
-  let eE' : Type max uE' uG uF uP := ULift E'
-  let eF : Type max uF uG uE' uP := ULift F
-  let eP : Type max uP uG uE' uF := ULift P
+    `contDiffOn_convolution_right_with_param_aux`. We reduce to this situation by pushing
+    everything through `ULift` continuous linear equivalences. -/
+  let eG : Type max uG uE' uF uP := ULift.{max uE' uF uP} G
+  -- porting note: `borelize eG` doesn't work
+  let _ : MeasurableSpace eG := borel eG; let _ : BorelSpace eG := ⟨rfl⟩
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/Convolution.lean:1486]: No, it seems that `set` works differently in Lean 4 and `borelize` doesn't support it.

File: Mathlib/Analysis/NormedSpace/OperatorNorm.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/NormedSpace/OperatorNorm.lean:619]: I renamed to `le_of_op_norm₂_le_of_le`. Feel free to rename again.

File: Mathlib/Analysis/Convolution.lean [hunk 5]
```diff
@@ -1210,56 +1165,57 @@ variable [NormedAddCommGroup G] [BorelSpace G]
...
-        (ball_subset_closed_ball hx)
-  · rw [integrable_indicator_iff hK'.measurable_set]
-    exact ((hf.integrable_on_is_compact hK').norm.const_mul _).mul_const _
-  · exact eventually_of_forall fun t x hx => (L _).HasFDerivAt.comp x (h3 x t)
+        (ball_subset_closedBall hx)
+  · rw [integrable_indicator_iff hK'.measurableSet]
+    exact ((hf.integrableOn_isCompact hK').norm.const_mul _).mul_const _
+  · exact eventually_of_forall fun t x _ => (L _).hasFDerivAt.comp x (h3 x t)
+  · exact hcg.convolutionExists_right L hf hg.continuous x₀
+-- · exact K'.indicator fun t => ‖L'‖ * ‖f t‖ * ⨆ x, ‖fderiv 𝕜 g x‖
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/Convolution.lean]: I guess, an edit artifact.

File: Mathlib/Analysis/Convolution.lean [hunk 6]
```diff
@@ -1310,18 +1265,10 @@ section WithParam
...
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
-/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+set_option maxHeartbeats 300000 in
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/Convolution.lean:1271]: @Ruben-VandeVelde added this line. Redirecting.

File: Mathlib/Analysis/Convolution.lean [hunk 7]
```diff
@@ -1393,103 +1337,105 @@ theorem hasFDerivAt_convolution_right_with_param {g : P → G → E'} {s : Set P
...
       apply add_subset_add
       · rw [neg_subset_neg]
-        apply closure_minimal (support_subset_iff'.2 fun z hz => _) hk.is_closed
+        refine closure_minimal (support_subset_iff'.2 fun z hz => ?_) hk.isClosed
         apply g'_zero x.1 z (h₀ε _) hz
         rw [mem_ball_iff_norm]
         exact ((le_max_left _ _).trans_lt hx).trans_le δε
-      · simp only [add_ball, thickening_singleton, zero_vadd]
+      · simp only [add_ball, thickening_singleton, zero_vadd, subset_rfl]
+    -- does not work while the above instance is not fixed
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/Convolution.lean:1392]: Merged `master`, waiting for the CI.

File: Mathlib/Analysis/Convolution.lean [hunk 8]
```diff
@@ -0,0 +1,1681 @@
...
+    (hg : IntegrableOn g (Ioi 0) μ) (L : E →L[ℝ] E' →L[ℝ] F) :
+    Integrable (posConvolution f g L ν) μ := by
+  rw [← integrable_indicator_iff (measurableSet_Ioi : MeasurableSet (Ioi (0 : ℝ)))] at hf hg
+  rw [posConvolution_eq_convolution_indicator f g L ν]
+  exact (hf.convolution_integrand L hg).integral_prod_left
+#align integrable_pos_convolution integrable_posConvolution
+
+/-- The integral over `Ioi 0` of a forward convolution of two functions is equal to the product
+of their integrals over this set. (Compare `integral_convolution` for the two-sided convolution.) -/
+theorem integral_pos_convolution [CompleteSpace E] [CompleteSpace E']
```
Comment (fpvandoorn) [inline Mathlib/Analysis/Convolution.lean:1669]:
```suggestion
theorem integral_posConvolution [CompleteSpace E] [CompleteSpace E']
```
(I know the statement doesn't actually mention `posConvolution`, though it should).

**Review by fpvandoorn — COMMENTED** [review]
LGTM, if all other review comments are also addressed.

File: Mathlib/Analysis/Convolution.lean (hunk 4 above)
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/Convolution.lean:1487]: Merged `master`, works now.

**Pull request closed without merging**